plugins/lua/buildingplan/unlink_mechanisms.lua

```diff
@@ -0,0 +1,678 @@
+                {
+                    "Really free mechanism? This cannot be undone.\n\nThis mechanism: ",
+                    {text = dfhack.items.getDescription(item, 0, true), pen = COLOR_BLUE},
+                },
+                desired_lines = 3,
+                propagate_fn = function(suppress)
+                    always_free_single = suppress
+                    item.flags.in_building = false
+                    dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(item)))
+                end,
+            }:show()
+        end
+    else
+        dfhack.printerr(("MechItemOverlay: Invalid button! Offset %d"):format(button.frame.t))
+    end
+end
+
+function MechItemOverlay:ask_free_all()
+    if always_free_all then
+        self:do_free_all()
+        return
+    end
+
+    ConfirmScreen
+    {
+        title = "Free All Mechanisms",
+        message = "Really free all mechanisms? This cannot be undone.",
+        desired_lines = 1,
+        propagate_fn = function(suppress)
+            always_free_all = suppress
+            self:do_free_all()
+        end,
+    }:show()
+end
+
+function MechItemOverlay:do_free_all()
+    for item in mech_iter(self.building) do
+        local target = get_mech_target(item)
+
+        if not target then
+            item.flags.in_building = false
+            dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(item)))
+        end
+    end
+end
+
+function MechItemOverlay:update_buttons()
+    self.subviews.free_all.visible = false
+
+    local scroll_pos = sheet.scroll_position_item
+    local bci_len = #self.building.contained_items
+    for i=1, self.num_buttons do
+        local button = self:get_button(i, true)
+        local offset = i*3 - 1 - ((scroll_pos + 1) % 3)
+        local idx = self:idx_from_offset(offset)
+
+        button.visible = false
+        if idx > 0 and idx < bci_len then
+            local item = self.building.contained_items[idx].item
+            if item._type == df.item_trappartsst and item.flags.in_building and
+                not item.flags.in_job and not get_trigger_index(item) then
+                    button.frame.t = offset
+                    button.visible = true
+                    self.subviews.free_all.visible = true
+            end
+        end
+        button:updateLayout()
+    end
+end
+
+function MechItemOverlay:fix_layout()
+    for i=1, self.num_buttons do --hide existing buttons
+        local button = self:get_button(i)
+        if button then
+            button.visible = false
+        end
+    end
+
+    self.num_buttons = (self.parent_height - 50) // 3
+    if not has_link_tab(self.building) then --space for extra button
+        self.num_buttons = self.num_buttons + 1
+    end
+
+    self.frame.h = self.num_buttons*3 + 2
+
+    if self.frame.h < 11 then --space between list and "free all" is removed
+        self.frame.h = self.frame.h - 1
+    end
+
+    self.frame.b = 4
+    self.frame.r = 40
```

Comment:
I haven't done a lot of testing on various screen resolutions. I used a 1600x900 window with `Interface scaling percentage` at 100%, 120%, 130%, and 140%. That's 4 data points, so I don't know if these values are precise.

Messing with `Maximum interface percentage` breaks horizontal layout. I don't think any of our overlays handle this properly. (Work order conditions overlay, for instance.)

Reply:
> I used a 1600x900 window with Interface scaling percentage at 100%, 120%, 130%, and 140%. That's 4 data points, so I don't know if these values are precise.

In my experience, we handle interface scaling without issues.

> Messing with Maximum interface percentage breaks horizontal layout. I don't think any of our overlays handle this properly.

very very true. this has been a known (to me) deficiency for a long time (since v50 introduced the concept of maximum interface percentage). I think we can solve it centrally, in `overlay.lua`, adjusting the position of the widgets according to the interface percentage.

Reply:
> > I used a 1600x900 window with Interface scaling percentage at 100%, 120%, 130%, and 140%. That's 4 data points, so I don't know if these values are precise.
> 
> In my experience, we handle interface scaling without issues.

I didn't know the extra space was going to disappear until I ran into it. I don't know that I can test arbitrary values of `parent_height`, so I can't be certain that the space doesn't disappear slightly earlier than `self.frame.h < 11`.

`(self.parent_height - 49) // 3` also fits my data points, but I don't know if that makes a difference.